pygmt/clib/session.py

```diff
@@ -2066,6 +2079,18 @@ def virtualfile_to_raster(
                 self["GMT_IS_IMAGE"]: "image",
                 self["GMT_IS_CUBE"]: "cube",
             }[family]
+
+        if kind == "image":  # Use temporary file for images
+            if not _HAS_RIOXARRAY:
+                raise ImportError(
+                    "Package `rioxarray` is required to be installed to load images. "
+                    "Please use `python -m pip install rioxarray` or "
+                    "`mamba install -c conda-forge rioxarray` to install the package."
+                )
+            with GMTTempFile(suffix=".tif") as tmpfile:
+                self.call_module("write", f"{vfname} {tmpfile.name} -Ti")
+                dataarray = rioxarray.open_rasterio(tmpfile.name)  # type: ignore[return-value]
+                return dataarray.load()
```

Comment:
Continuing from https://github.com/GenericMappingTools/pygmt/issues/3315#issuecomment-2212420503. This part is causing an error on Windows:

```python-traceback
        path = which("@earth_day_01d", download="a")
        raster = rioxarray.open_rasterio(path)
>       result = grdcut(raster, region=region)

..\pygmt\tests\test_grdcut.py:133: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\pygmt\helpers\decorators.py:609: in new_module
    return module_func(*args, **kwargs)
..\pygmt\helpers\decorators.py:773: in new_module
    return module_func(*bound.args, **bound.kwargs)
..\pygmt\src\grdcut.py:127: in grdcut
    return lib.virtualfile_to_raster(
..\pygmt\clib\session.py:2090: in virtualfile_to_raster
    with GMTTempFile(suffix=".tif") as tmpfile:
..\pygmt\helpers\tempfile.py:77: in __exit__
    Path(self.name).unlink(missing_ok=True)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = WindowsPath('C:/Users/RUNNER~1/AppData/Local/Temp/pygmt-nwxyygd3.tif')
missing_ok = True

    def unlink(self, missing_ok=False):
        """
        Remove this file or link.
        If the path is a directory, use rmdir() instead.
        """
        try:
>           os.unlink(self)
E           PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\RUNNER~1\\AppData\\Local\\Temp\\pygmt-nwxyygd3.tif'
```

Can you try returning the dataarray fully loaded into memory like so:

```suggestion
            with GMTTempFile(suffix=".tif") as tmpfile:
                self.call_module("write", f"{vfname} {tmpfile.name} -Ti")
                dataarray = rioxarray.open_rasterio(tmpfile.name).load()
                return dataarray
```

Reply:
```
            with GMTTempFile(suffix=".tif") as tmpfile:
                self.call_module("write", f"{vfname} {tmpfile.name} -Ti")
                with rioxarray.open_rasterio(tmpfile.name) as da:
                    dataarray = da.load()
                return dataarray
```
This one works. I guess we just need to ignore the mypy errors.

Reply:
That `Item "list[Dataset]" of "Dataset | DataArray | list[Dataset]" has no attribute "__enter__"  [union-attr]` mypy error seems to be an upstream bug. Probably should report it, but I'm unsure whether to post it to rioxarray or xarray...

Reply:
https://corteva.github.io/rioxarray/html/_modules/rioxarray/_io.html#open_rasterio

I guess it's because `open_rasterio` is defined as a normal function rather than a context manager. It's unclear why `with rioxarray.open_rasterio(tmpfile.name) as da:` works, but as you mentioned above, the rioxarray IO is really complicated.